### ISSUE TITLE
method = 'predict' is not working

### ISSUE THREAD

**Issue by snilsn** [tobac-project/tobac#169]
Another bug related to the linking process:
`method=predict` of `linking_trackpy()` is not working at the moment.

It's a problem of `trackpy`, I have documented my findings in a comment here: 

https://github.com/soft-matter/trackpy/issues/699

I think we have three options:

1. Wait for a fix in trackpy. That might take a while though, the issue hasn`t been touched in 2 months
2. Fix the problem in trackpy ourselves
3. Implement the workaround I found with `link_df_iter` in tobac

I would probably go with the third option.

**Comment by freemansw1** [comment]
This explains some of the strange behavior I have noticed recently. Yes, we should fix this ASAP. I'm happy for us to implement a workaround; what is the workaround that you have proposed?

**Comment by freemansw1** [comment]
Long term, as you note, we should probably work on a fix upstream. They were pretty receptive to my last PR, but it did take a while to work through.

**Comment by snilsn** [comment]
> what is the workaround that you have proposed?

The implementation in `tobac` looks like this:

```
pred = tp.predict.NearestVelocityPredict(span=1)
trajectories_unfiltered = pred.link_df(
            features_linking,
            search_range=search_range,
            memory=memory,
            pos_columns=["hdim_1", "hdim_2"],
            t_column="frame",
            neighbor_strategy="KDTree",
            link_strategy="auto",
            adaptive_step=adaptive_step,
            adaptive_stop=adaptive_stop
        )
```
We could avoid the usage of `link_df` by generating an iterable of the frames and use 'link_df_iter' instead:
```
features_linking_list = [frame for i, frame in features.groupby('frame')]
pred = tp.predict.NearestVelocityPredict(span=1)
trajectories_unfiltered = pred.link_df_iter(
            features_linking_list,
            search_range=search_range,
            memory=memory,
            pos_columns=["hdim_1", "hdim_2"],
            t_column="frame",
            neighbor_strategy="KDTree",
            link_strategy="auto",
            adaptive_step=adaptive_step,
            adaptive_stop=adaptive_stop
trajectories_unfiltered = pd.concat(trajectories_unfiltered)
```
There probably are smarter and more efficient ways to do this, that's just how I got it to actually predict something quick and dirty.

**Comment by snilsn** [comment]
> Long term, as you note, we should probably work on a fix upstream.

Yes, and the issue I commented on in trackpy already proposes something, but that didn't fix it for me.

>  They were pretty receptive to my last PR, but it did take a while to work through.

Yup, in a different issue I opened there I got a really quick answer. I probably should have opened a separate issue for this problem instead of using the old one.

**Comment by freemansw1** [comment]
I would be happy for you to submit this as a PR if you are so inclined. I'm curious how much worse performance is with the fix; if you submit as a PR I can test on my huge dataset.

**Comment by freemansw1** [comment]
Also, yeah, I'd open a new issue upstream with your very detailed thoughts. Trackpy gets a lot of issues, and I suspect that that one may have been missed.

**Comment by snilsn** [comment]
Oh, sorry, I got something mixed up. 
It is actually necessary to change `trackpy` as it is suggested [here](https://github.com/soft-matter/trackpy/issues/699#issue-1261383739).
The workaround alone wouldn't fix the problem, unfortunately.
I got confused with my environments.

**Comment by freemansw1** [comment]
Does the workaround proposed in that issue actually fix the problem? If so, I think one of us should submit it as a PR to get their attention.

**Comment by snilsn** [comment]
For anyone interested, here is my basic example:

```
import trackpy as tp
import pandas as pd
import matplotlib.pyplot as plt

d = {'frame': [1, 2, 3], 'x': [0, 1, 3], 'y': [0, 1, 3]}
df = pd.DataFrame(data = d)

fig, ax = plt.subplots(ncols=3, sharey=True, figsize=(15, 5))

track = tp.link(df, 2)
ax[0].set_title('Without prediction')
tp.plot_traj(track, plot_style={'marker':'d'}, ax=ax[0])

pred = tp.predict.NearestVelocityPredict()
track = pred.link_df(df, 2)
ax[1].set_title('Prediction with link_df')
tp.plot_traj(track, plot_style={'marker':'d'}, ax=ax[1])

pred = tp.predict.NearestVelocityPredict()
track = pred.link_df_iter([frame for i,frame in df.groupby('frame')], 2)
ax[2].set_title('Prediction with link_df_iter')
tp.plot_traj(pd.concat(track), plot_style={'marker':'d'}, ax=ax[2])

plt.savefig('variants.jpeg')
```


![variants](https://user-images.githubusercontent.com/76663232/185658349-5f3caf42-af2b-4c1d-85c5-a1190e16e124.jpeg)

**Comment by freemansw1** [comment]
I'm closing this as it has been fixed by #170 in the hotfix branch.
